File: Python/83.Remove Duplicates from Sorted List.py
#有序链表中的数字重复的结点只保留一个
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        if head==None:
            return head
        p = head
        while p.next!=None:
            if p.val == p.next.val:
                p.next = p.next.next
            else:
                p = p.next
        return head



l1 = ListNode(1)
l2 = ListNode(1)
l3 = ListNode(2)
l1.next = l2
l2.next = l3

l4 = ListNode(5)
l5 = ListNode(6)
l6 = ListNode(4)
l4.next = l5
l5.next = l6

l7 = ListNode(5)
l8 = ListNode(5)

l9 = ListNode(9)
l10 = ListNode(9)

s = Solution()
p = s.deleteDuplicates(l1)
while p!=None:
    print(p.val)
    p = p.next

File: C++/Linked List/83.Remove Duplicates from Sorted List.cpp
/*
删除有序链表中数字重复的结点。
*/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* deleteDuplicates(ListNode* head) {
		if(!head)	return NULL;
		ListNode *cur=head;
        while(cur->next)
        {
        	if(cur->val == cur->next->val)
        		cur->next = cur->next->next;
        	else cur = cur->next;
        }
        return head;

        /*递归写法
        if(head == NULL || head->next == NULL)return head;
        head->next = deleteDuplicates(head->next);
        return head->val == head->next->val ? head->next : head;
         */
    }
};
int main()
{
	Solution s;
	ListNode head(1);
	ListNode t1(2);
	ListNode t2(2);
	ListNode t3(4);
	ListNode t4(5);
	ListNode t5(5);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.deleteDuplicates(&head);
 	while(h)
 	{
 		cout<<h->val<<endl;
 		h = h->next;
 	}
}

